datafusion/common/src/scalar.rs

```diff
@@ -1368,103 +1366,36 @@ impl ScalarValue {
-                        sv => _internal_err!(
-                            "Inconsistent types in ScalarValue::iter_to_array. \
-                                Expected {:?}, got {:?}",
-                            data_type, sv
-                        ),
-                    })
-                    .collect::<Result<Vec<_>>>()?,
-                )))
-            }};
-        }
-
-        macro_rules! build_array_list_string {
-            ($BUILDER:ident, $STRING_ARRAY:ident,$LIST_BUILDER:ident,$SCALAR_LIST:pat) => {{
-                let mut builder = $LIST_BUILDER::new($BUILDER::new());
-                for scalar in scalars.into_iter() {
-                    match scalar {
-                        ScalarValue::List(arr) if matches!(scalar, $SCALAR_LIST) => {
-                            // `ScalarValue::List` contains a single element `ListArray`.
-                            let list_arr = as_list_array(&arr);
-
-                            if list_arr.is_null(0) {
-                                builder.append(false);
-                                continue;
-                            }
-
-                            let string_arr = $STRING_ARRAY(list_arr.values());
-
-                            for v in string_arr.iter() {
-                                if let Some(v) = v {
-                                    builder.values().append_value(v);
-                                } else {
-                                    builder.values().append_null();
-                                }
-                            }
-                            builder.append(true);
-                        }
-                        ScalarValue::LargeList(arr) if matches!(scalar, $SCALAR_LIST) => {
-                            // `ScalarValue::List` contains a single element `ListArray`.
-                            let list_arr = as_large_list_array(&arr);
-
-                            if list_arr.is_null(0) {
-                                builder.append(false);
-                                continue;
-                            }
-
-                            let string_arr = $STRING_ARRAY(list_arr.values());
-
-                            for v in string_arr.iter() {
-                                if let Some(v) = v {
-                                    builder.values().append_value(v);
-                                } else {
-                                    builder.values().append_null();
-                                }
-                            }
-                            builder.append(true);
-                        }
-                        sv => {
-                            return _internal_err!(
-                                "Inconsistent types in ScalarValue::iter_to_array. \
-                                    Expected List, got {:?}",
-                                sv
-                            )
-                        }
-                    }
+        fn build_list_array(
```

Comment:
😍 Amazing @jayzhan211 

I believe as written, this function will always return `ListArray` (never `LargeListArray`) so I think this probably needs to be templated on `<O: OffsetSizeTrait>` -- the way @Weijun-H  has done in https://github.com/apache/arrow-datafusion/pull/8322 for example)

Reply:
I think mutable handle large list too, otherwise the large list test should fail 🤔

Reply:
You are right